ts/util/blockedNumberController.ts

```diff
@@ -0,0 +1,124 @@
+import { createOrUpdateItem, getItemById } from '../../js/modules/data';
+import { PubKey } from '../session/types';
+import { MultiDeviceProtocol } from '../session/protocols';
+
+const BLOCKED_NUMBERS_ID = 'blocked';
+const BLOCKED_GROUPS_ID = 'blocked-groups';
+
+// tslint:disable-next-line: no-unnecessary-class
+export class BlockedNumberController {
+  private static loaded: boolean = false;
+  private static blockedNumbers: Set<string> = new Set();
+  private static blockedGroups: Set<string> = new Set();
+
+  /**
+   * Check if a device is blocked.
+   * Make sure `load()` has been called before this function so that the correct blocked state is returned.
+   *
+   * @param number The device.
+   */
+  public static isBlocked(device: string | PubKey): boolean {
+    // This function is not `async` because the old `isBlocked` function in js was also not async.
+    // To convert it means we'll have to re-wire all our UI components to work with async.
+    const stringValue =
+      device instanceof PubKey ? device.key : device.toLowerCase();
+    return this.blockedNumbers.has(stringValue);
+  }
+
+  /**
+   * Check if a group id is blocked.
+   * Make sure `load()` has been called before this function so that the correct blocked state is returned.
+   *
+   * @param groupId The group id.
+   */
+  public static isGroupBlocked(groupId: string | PubKey): boolean {
+    // This function is not `async` because the old `isBlocked` function in js was also not async.
+    // To convert it means we'll have to re-wire all our UI components to work with async.
+    const stringValue =
+      groupId instanceof PubKey ? groupId.key : groupId.toLowerCase();
+    return this.blockedGroups.has(stringValue);
+  }
+
+  /**
+   * Block a user (including their linked devices).
+   *
+   * @param user The user to block.
+   */
+  public static async block(user: string | PubKey): Promise<void> {
+    // The reason we add all linked device to block number set instead of checking if any device of a user is in the `isBlocked` function because
```

Comment:
This would be a problem if the user adds a device, right?

Reply:
It would. Though i have some fixes we can apply on the receiving end and in conversation which might help out.

Reply:
Fix in my new commit